File: Solutions/Python/generate_parentheses.py
"""
    [MEDIUM]
    22. Generate Parentheses

    Concepts:
    - backtracking
    - depth-first search

    Stats:
        Runtime | 3 ms      [Beats 31.45%]
        Memory  | 17.93 MB  [Beats 59.15%]
"""

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        """
            If we map each parenthesis to a move in a 
            matrix, problem becomes path enumeration from DFS
        """
        paths = []

        def dfs(i, j, path):
            if (i, j) == (n, n):
                paths.append("".join(path))

            if i <= n:
                dfs(i + 1, j, path + ["("])
            if i > j and j <= n:
                dfs(i, j + 1, path + [")"])

        dfs(0, 0, [])
        return paths

File: Solutions/Cpp/generate_parentheses.cpp
/*
    [MEDIUM]
    22. Generate Parentheses

    Concepts:
    - backtracking
    - depth-first search

    Stats:
        Runtime | 11 ms     [Beats 8.46%]
        Memory  | 16.48 MB  [Beats 14.33%]
*/

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        /*
            If we map each parenthesis to a move in a 
            matrix, problem becomes path enumeration from DFS
        */
        std::vector<std::string> paths = {};

        std::function<void(int, int, std::string)> dfs;
        dfs = [&](int i, int j, std::string path){
            if (i == n && j == n)
                paths.push_back(path);

            if (i <= n)
                dfs(i + 1, j, path + "(");
            if (i > j && j <= n)
                dfs(i, j + 1, path + ")");
        };

        dfs(0, 0, "");
        return paths;
    }
};